src/binary/writer.rs

```diff
@@ -868,6 +860,76 @@ impl<W: Write> BinarySystemWriter<W> {
     }
 }
 
+/// These functions provide "raw" encoding steps of values into Ion binary.
+///
+/// For example, consider an i64 as defined in [the spec][spec]:
+///
+/// ```text
+///            7       4 3       0
+///           +---------+---------+
+/// Int value |  2 or 3 |    L    |
+///           +---------+---------+======+
+///           :     length [VarUInt]     :
+///           +==========================+
+///           :     magnitude [UInt]     :
+///           +==========================+
+/// ```
+///
+/// There are 3 steps here:
+///
+/// 1. The type (positive or negative int)
+/// 2. The length of the binary representation of the int
+/// 3. The magnitude of the int
+///
+/// To represent an i64 in Ion binary, all 3 steps need to be taken. However, to
+/// include an i64 in an Ion hash computation, the length is not relevant!
+/// That's why this factoring is useful: it enables code sharing of the
+/// individual steps.
+///
+/// [spec]: https://amzn.github.io/ion-docs/docs/binary.html#2-and-3-int
+pub mod raw {
+    use super::*;
+
+    /// The result of `i64_magnitude`. The fields are private; users must go
+    /// through `as_ref()` to see the slice that ignores leading zeros.
+    ///
+    /// This type exists to avoid heap allocation.
+    pub struct I64Magnitude {
+        magnitude_bytes: [u8; mem::size_of::<u64>()],
+        first_occupied_byte: usize,
+    }
+
+    impl AsRef<[u8]> for I64Magnitude {
+        fn as_ref(&self) -> &[u8] {
+            &self.magnitude_bytes[self.first_occupied_byte..]
+        }
+    }
```

Comment:
I agree with the point around autocomplete, but *in my opinion* I don't think it's as important as being idiomatic. Now, I'm not stating that what I've done **is** idiomatic - rather, that I **think** I am being idiomatic! Part of the benefit of being idiomatic is the tooling will catch up to the idioms.

I'm fine with having two methods (one being the idiomatic one, the other being the one that works with autocomplete), but honestly I don't think it's worth it - especially not for this "low level" stuff that we really don't expect the majority of users to use.

Reply:
I don't have a strong opinion, but if I think of `I64Magnitude` as a slice, then the `AsRef<[u8]>` makes sense because it is essentially a container for producing that slice reference.

That said, I do like also having an explicit method that gives the option to make the call site more obvious when not using the trait, this is a common pattern in a lot of types in the Rust library but those are typically in types that implement `AsRef<...>` more than once (e.g. `String` that has `as_str()`/`as_bytes()` and implements `AsRef<str>` and `AsRef<[u8]>`.

Reply:
> this is a common pattern in a lot of types in the Rust library but those are typically in types that implement AsRef<...> more than once (e.g. String that has as_str()/as_bytes() and implements AsRef<str> and AsRef<[u8]>.

This was exactly the example I had in mind, thanks for adding this. I should've been more explicit.